pkg/segment/aggregations/segaggs.go

```diff
@@ -2408,6 +2414,665 @@ func performValueColRequest(nodeResult *structs.NodeResult, aggs *structs.QueryA
+func getFloatValForBin(fieldToValue map[string]segutils.CValueEnclosure, record map[string]interface{}, field string) (float64, error) {
+	var err error
+	var fieldValue interface{}
+	var exist bool
+	if record != nil {
+		fieldValue, exist = record[field]
+		if !exist {
+			return 0, fmt.Errorf("getFloatValForBin: field %s does not exist in record", field)
+		}
+	} else {
+		fieldCValue, exist := fieldToValue[field]
+		if !exist {
+			return 0, fmt.Errorf("getFloatValForBin: field %s does not exist in record", field)
+		}
+		fieldValue = fieldCValue.CVal
+	}
+
+	fieldValueFloat, isNumeric := getNumericValue(fieldValue)
+	if !isNumeric {
+		return 0, fmt.Errorf("getFloatValForBin: field %s is not a numeric, err: %v", field, err)
+	}
+
+	return fieldValueFloat, nil
+}
+
+// Function to find the span range length
+func findSpan(minValue float64, maxValue float64, maxBins uint64, minSpan *structs.BinSpanLength, field string) *structs.BinSpanOptions {
+	if field == "timestamp" {
+		return findEstimatedTimeSpan(minValue, maxValue, maxBins, minSpan)
+	}
+	if minValue == maxValue {
+		return &structs.BinSpanOptions{
+			BinSpanLength: &structs.BinSpanLength{
+				Num:       1,
+				TimeScale: segutils.TMInvalid,
+			},
+		}
+	}
+
+	// span ranges estimated are in powers of 10
+	span := (maxValue - minValue) / float64(maxBins)
+	exponent := math.Log10(span)
+	exponent = math.Ceil(exponent)
+	spanRange := math.Pow(10, exponent)
+
+	// verify if estimated span gives correct number of bins, refer the edge case like 301-500 for bins = 2
+	for {
+		lowerBound, _ := getBinRange(minValue, spanRange)
+		_, upperBound := getBinRange(maxValue, spanRange)
+
+		if (upperBound-lowerBound)/spanRange > float64(maxBins) && spanRange <= math.MaxFloat64/10 {
+			spanRange = spanRange * 10
+		} else {
+			break
+		}
+	}
+
+	// increase the spanRange till minSpan is satisfied
+	if minSpan != nil {
+		for {
+			if spanRange < minSpan.Num && spanRange <= math.MaxFloat64/10 {
+				spanRange = spanRange * 10
+			} else {
+				break
+			}
+		}
+	}
+
+	return &structs.BinSpanOptions{
+		BinSpanLength: &structs.BinSpanLength{
+			Num:       spanRange,
+			TimeScale: segutils.TMInvalid,
+		},
+	}
+}
+
+// Function to bin ranges with the given span length
+func getBinRange(val float64, spanRange float64) (float64, float64) {
+	lowerbound := math.Floor(val/spanRange) * spanRange
+	upperbound := math.Ceil(val/spanRange) * spanRange
+	if lowerbound == upperbound {
+		upperbound += spanRange
+	}
+
+	return lowerbound, upperbound
+}
+
+func getSecsFromMinSpan(minSpan *structs.BinSpanLength) float64 {
+	if minSpan == nil {
+		return 0
+	}
+
+	switch minSpan.TimeScale {
+	case segutils.TMSecond:
+		return minSpan.Num
+	case segutils.TMMinute:
+		return minSpan.Num * 60
+	case segutils.TMHour:
+		return minSpan.Num * 3600
+	case segutils.TMDay:
+		return minSpan.Num * 86400
+	default:
+		// default returning num*(seconds in a month)
+		return minSpan.Num * 2592000
+	}
```

Comment:
No, the default value should be months when timescale is larger than a day like week, month, quarter or year. I am just returning a larger number which would be give us the estimated span in months.

Reply:
Ok, then you should handle those cases explicitly like:
```
case segutils.TMWeek, segutils.TMMonth, segutils.TMQuarter, ...:
    // do what you did in the default case
default:
    // log and return an error
```
This will help us detect errors faster (e.g., what happens for `TMMicroseconds`? It probably shouldn't be treated the same as `TMMonth`, right?), and if we ever add a new enum value (e.g., `TMNanosecond`) we'd be able to detect the issue from the log messages

Reply:
addressed in the new commit.